File: app/services/audit.py
from __future__ import annotations

from typing import Any, Dict

from sqlalchemy.orm import Session

from app.models import AuditLog, User


def log_action(db: Session, *, user: User, action: str, meta: Dict[str, Any] | None = None) -> AuditLog:
    entry = AuditLog(user_id=user.id, action=action, meta_json=meta or {})
    db.add(entry)
    db.commit()
    db.refresh(entry)
    return entry

File: server/src/services/audit.ts
import type { AuditLog } from "@prisma/client";
import { prisma } from "../db.js";

export async function logAction({
  userId,
  action,
  meta,
}: {
  userId: number;
  action: string;
  meta?: Record<string, unknown>;
}): Promise<AuditLog> {
  return prisma.auditLog.create({
    data: {
      userId,
      action,
      metaJson: meta ?? {},
    },
  });
}

